Wireguard VPN support

jukkar opened the pull request:
This is initial Wireguard VPN support. Part of the implementation is ported from wireguard-lwip project.
Some discussion about this can be found in https://github.com/zephyrproject-rtos/zephyr/discussions/63722

Some of the commits in this PR might be sent separately to review.

jukkar committed:
net: wireguard: Add initial implementation

Wireguard is a VPN implementation. See https://www.wireguard.com/
for details.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: virtual: Add capability flag for VPN type interface

Allow user to mark the virtual interface as a VPN one.
This is used by Wireguard implementation.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: Add Wireguard VPN support

Add commands to manipulate Wireguard connectivity.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: Add crypto support

The crypto code is taken from wireguard-lwip project at
https://github.com/smartalock/wireguard-lwip
and is BSD-3-Clause and MIT licensed code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: crypto: Add SPDX license identifiers

Add relevant SPDX license identifier to crypto algorithm
implementations.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: l2: virtual: Add support for VPN public/private key set/get

Add support for getting public address and setting private
key for the virtual interface. This is needed for Wireguard VPN.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: iface: Print Wireguard public key

If the interface is Wireguard VPN interface, then print
the public key of the interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Add helper to get src interface and address from dst address

Instead of calling various network interface API functions to get
the network interface and related source IP address, have a single
function that can return both data.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-server: Add Wireguard VPN support

Add Wireguard configuration to echo-server application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: Add helper to get sockaddr from sockaddr_storage struct

As one should not instantiate sockaddr struct (only have pointers
to it), but only use struct sockaddr_storage to store the
socket address information. In order to simplify the casting,
create net_sad() macro that casts a "struct sockaddr_storage *" to
"struct sockaddr *" which can then be used by other macros in the
networking stack.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: crypto: Fix CI compliance issues

Make the Wireguard crypto code pass Zephyr CI compliance check script.
Following changes were done:

* Convert // comments to /* */ comments
* Run the code through clang-format
* Empty line after variable declaration
* Removing Tab after Space characters
* Removing trailing white space
* Removing empty lines at the end of the file
* Changed "unsigned i" to "unsigned int i"

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: crypto: Fix warning in x25519.c mul()

Compiler gives this warning

In function ‘ladder_part1’,
    inlined from ‘x25519_core’ at wireguard/crypto/refc/x25519.c:306:3,
    inlined from ‘x25519’ at wireguard/crypto/refc/x25519.c:317:2:
wireguard/crypto/refc/x25519.c:259:9: warning: ‘mul’ reading 32 bytes
 from a region of size 4 [-Wstringop-overread]
  259 |         mul(z2, x2, a24, ARRAY_SIZE(a24));  /*  z2 = E*a24 */
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
wireguard/crypto/refc/x25519.c: In function ‘x25519’:
wireguard/crypto/refc/x25519.c:259:9: note: referencing argument 3 of
 type ‘const limb_t *’ {aka ‘const unsigned int *’}
wireguard/crypto/refc/x25519.c:148:13: note: in a call to function ‘mul’
  148 | static void mul(fe out, const fe a, const fe b, unsigned int nb)
      |             ^~~

Fix for this can be found from original sources at

https://sourceforge.net/p/strobe/code/ci/0aa9e2abcaa4e6364c97a914d397517668475209/

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-client: Add Wireguard VPN support

Add Wireguard configuration to echo-client application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: echo-client: Enable event mgmt info support

CONFIG_NET_MGMT_EVENT_INFO needs to be enabled for this sample so
that we can get detailed information when the event is generated.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: http-server: Add Wireguard VPN support

Add Wireguard configuration to http-server application.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Updated to latest main
* Fixed compilation issues
* Added VPN support to echo-client and http-server sample applications

jukkar committed:
samples: net: Enable Wireguard VPN compilation in tests

Add a test that enables Wireguard VPN compilation so that
we at least compile test the code.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: stats: Add statistics support

Collect Wireguard VPN statistics and allow user to fetch it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: wg: Add VPN statistics support

Show VPN statistics support if enabled.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* manifest update for net-tools
* allowed ip address list fixes
* shell print allowed ips and endpoint
* more statistics collection

jukkar commented:
* housekeeping timer works now
* keepalive timer is working and can be set from application

jukkar commented:
* updated to latest main
* fixed Zephyr initiated handshake issues
* CI fixes

jukkar commented:
* updated to latest main
* compliance check fix
* refactored VLAN support in echo-server and echo-client samples in order to test VPN over VLAN use case

jukkar committed:
net: wireguard: Send network events for VPN activity

Send peer add/del network event when the peers is either added
to the system or deleted from the system.
Send VPN connected / disconnected event when a VPN connection
is successfully established or the peer connection is disconnected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: shell: events: Print wireguard event information

Add Wireguard VPN events information printouts to event monitor.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* convert other samples to use the common VLAN init (to make CI pass)
* let user select the wireguard port
* send net_mgmt events for when the VPN connection is established or disconnected, and when a new VPN peer is added or removed

jukkar committed:
samples: net: common: Add information about VPN and VLAN

Add example and information how to run VPN over a VLAN with
the echo-server sample.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Add information how to run echo-server with VPN running over VLAN

jukkar committed:
net: shell: wg: Print detailed peer information

The "net wg show 1" will show detailed information of the peer
id 1. This is useful when debugging connectivity issues.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Fixed timer when deciding if the connection init should be done
* Added support for `net wg show <id>` command which shows detailed information about the peer connection

pdgendt commented on include/zephyr/net/net_pkt.h:
```diff
@@ -123,6 +123,17 @@ struct net_pkt {
 #endif
 
+#if defined(CONFIG_WIREGUARD)
+	/* Original network interface */
+	struct net_if *wg_iface;
+	/* Pointer to IP header of the encrypted pkt */
+	union net_ip_header wg_ip_hdr;
+	/* Pointer to UDP header of the encrypted pkt */
+	union net_proto_header wg_proto_hdr;
+	/* Peer id */
+	int wg_peer_id;
```
Instead of prefixing all with `wg_`, put it in a struct?

```c
	struct {
		/* Original network interface */
		struct net_if *iface;
		/* Pointer to IP header of the encrypted pkt */
		union net_ip_header ip_hdr;
		/* Pointer to UDP header of the encrypted pkt */
		union net_proto_header proto_hdr;
		/* Peer id */
		int peer_id;
	} wg;
```

jukkar replied:
Sure, we can certainly do that.

pdgendt reviewed (commented):
There are a few fixup commits for the copied BSD-3 and MIT files, should these be squashed instead?

jukkar commented:
> There are a few fixup commits for the copied BSD-3 and MIT files, should these be squashed instead?

I considered it but then it is difficult to make a diff to the original files to see if there are any changes. We need to modify the source files from wireguard-lwip in order to pass compliance / code formatting check, and the fixup commit is for that. I think the fixup commit is a good compromise here. Readability of the original crypto commits is poor but that is "fixed" in the subsequent commits.

jukkar committed:
net: if: Add special handling for IPv4/6 address check for VPN

This is a hack that is used until we have proper IP routing
in place. The code has now special check that makes sure that
we only route IP packets to VPN interface when the packet is
destined to that subnet. So if destination IP address does
not belong to VPN interface subnet, it is not routed there.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Rebased to latest main
* Fixed allowed_ip list handling
* Updated according to comments
* Added a hack to not route general IP traffic to VPN interface (only traffic that belongs to VPN should go to the VPN interface). This quite hackish but I did not see better alternative for this until we have proper IP routing implemented. See the commit at the end of the patch train.

I am truly sorry about the amount of commits here, the amount of commits is simply very large. Fortunately most of them are quite small but I did not see a good alternative for the amount of patches. They are needed in order to be able to test the beast. :bow:

fabiobaltieri commented on samples/net/sockets/echo_server/overlay-vpn.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_WIREGUARD=y
+CONFIG_WIREGUARD_MAX_SRC_IPS=4
+
```
this really seems to want some more IP and stack, should these have some combo of

```
CONFIG_NET_IF_UNICAST_IPV6_ADDR_COUNT=4
CONFIG_NET_IF_UNICAST_IPV4_ADDR_COUNT=2
CONFIG_NET_IF_MAX_IPV6_COUNT=4
CONFIG_NET_IF_MAX_IPV4_COUNT=4

CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE=2048
CONFIG_NET_RX_STACK_SIZE=2048
```
?

jukkar replied:
Yes noticed that too, I will add relevant options there.

fabiobaltieri commented on subsys/net/lib/wireguard/wg.c:
```diff
@@ -0,0 +1,2078 @@
+	}
+
+	if (hdr->type == MESSAGE_HANDSHAKE_INITIATION) {
+		if (!IS_ENABLED(CONFIG_WIREGUARD_SERVER)) {
+			NET_DBG("DROP: %s not supported in client mode",
+				"handshake init");
+			goto drop;
+		}
```
is the protocol happy with this option? trying to run it with `CONFIG_WIREGUARD_SERVER=n`, seems like the peer really likes to init on their side sometimes, even after the connection has been established I occasionally get some

```
[00:03:55.173,000] <dbg> wireguard: wg_ctrl_recv: (rx_q[0]): DROP: handshake init not supported in client mode
```

and no traffic flows anymore, peer sees an increasing `latest handshake` in `wg` and that's it

fabiobaltieri replied:
Yeah something's off, with this enabled and some constant traffic from the peer the connection is stable, as soon as that stops data stops flowing after a while until the peer re-initiates

jukkar replied:
There is some issue when zephyr tries to connect to peer, I am investigating it.

jukkar replied:
> trying to run it with `CONFIG_WIREGUARD_SERVER=n`

That is a good question if we should have that server option. My motivation for it was to restrict the incoming traffic if the zephyr device is only suppose to connect out. And one way to support that was not to allow incoming init if server is not enabled, not sure if there is better alternative for that, or if we really need the server option at all.

fabiobaltieri commented on samples/net/sockets/echo_server/overlay-vpn.conf:
```diff
@@ -0,0 +1,8 @@
+CONFIG_WIREGUARD_MAX_SRC_IPS=4
+
+# First VPN interface will use these settings
+CONFIG_NET_SAMPLE_COMMON_VPN_PEER_IP_ADDR="192.0.2.2"
```
maybe add a port? all samples in https://www.wireguard.com/quickstart/ specify it too, seems like there's a bit of a hint of not sticking to the default one

jukkar replied:
That setting should already support the port, so you can say here
```
CONFIG_NET_SAMPLE_COMMON_VPN_PEER_IP_ADDR="192.0.2.2:51820"
```

fabiobaltieri replied:
yeah my thought was to specify it explicitly in the sample to match the official samples too, to make it obvious that it's supported

fabiobaltieri replied:
actually wg does not even have a default:

```
# wg set wg0 peer ... endpoint 1.2.3.4
Unable to find port of endpoint: `1.2.3.4'
```

jukkar replied:
I added the port number to couple of vpn overlay config file in samples.

fabiobaltieri commented on subsys/net/lib/shell/wg.c:
```diff
@@ -0,0 +1,392 @@
+			    &olen, peer->key.public_key,
+			    sizeof(peer->key.public_key));
+
+	PR("[%2d] %d     %-40s\t %s\t",
+	   peer->id, net_if_get_by_iface(peer->iface), public_key,
+	   net_addr_ntop(peer->cfg_endpoint.ss_family,
+			 &net_sin(net_sad(&peer->cfg_endpoint))->sin_addr,
+			 addr, sizeof(addr)));
```
print the endpoint port as well?

jukkar replied:
Good point, will add that.

jukkar commented:
* updated to latest main
* adjusted according to comments
* removed the `CONFIG_WIREGUARD_SERVER` option to be in line with Linux
* I was not able to fully solve the issue where zephyr is sending handshake init to Linux which drops them with this message `wireguard: zwg0: Invalid handshake initiation from 192.168.88.27:51820`. Zephyr sends the init messages multiple times and eventually the connection succeeds but it is a bit mystery still why the messages are dropped by Linux. Fortunately the situation seems to solve itself after some time (which could be somewhat longish, I saw cases that it took about 2 min before the connection started to work). I am suspecting the `greatest_timestamp` field in the peer struct which should increase over time, but if Zephyr boots itself, then the value is reset. But this is speculation atm, and that field should probably be made persistent but it is a bit difficult atm with the network samples.

fabiobaltieri commented on subsys/net/ip/Kconfig:
```diff
@@ -323,6 +324,11 @@ config NET_SHELL_WEBSOCKET_SUPPORTED
 	depends on NET_SHELL_SHOW_DISABLED_COMMANDS || WEBSOCKET_CLIENT
 
+config NET_SHELL_WIREGUARD_SUPPORTED
+	bool "Wireguard VPN"
```
should it say experimental for the moment?

jukkar committed:
net: ipv4: Add DiffServ DSCP code point definitions

Wireguard packets should be classified so we need these
DSCP defines.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wireguard: Add DiffServ DSCP values

As described in https://www.wireguard.com/protocol/
set DSCP to AF41 for handshake messages. The transport data
DSCP is set to 0.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Fixed the timeout checks when deciding when to send handshake (stupid mistake where I had mixed ms and sec :facepalm: ) This should fix the connection initiation and periodic handshake issue and the VPN connection seems to stay up now properly.
* Added DiffServ DSCP code setting to IPv4 packets as described in Wireguard protocol page https://www.wireguard.com/protocol/

rlubos reviewed (dismissed):
It's just too massive to review in detail, but the integration part looks neat and the common code extraction from samples make sense. +1 from my side

fabiobaltieri commented on subsys/net/lib/wireguard/wg_crypto.c:
```diff
@@ -0,0 +1,1344 @@
+	 * 64 bit seconds from 1970 = 8 bytes
+	 * 32 bit nano seconds from current second
+	 */
+	uint64_t millis = k_ticks_to_ms_floor64(sys_clock_tick_get());
```
Hey played a bit more with this, tested it both on an STM32H755 and an ESP32C3, just to introduce some variations. :-) Seems to work great in the current configuration but it seems like it has a hard time reconnecting after a reset, and reconfiguring the remote interface or waiting makes it work immediately. Searched a bit and I bumped into this: https://www.reddit.com/r/WireGuard/comments/q47pss/client_wont_reconnect_until_server_restarted/, quoting

> Timestamps are part of the Wireguard handshake. Wireguard remembers the timestamps the peers have sent the last handshake and requires them to increase monotonically as a way to fight replay attacks. If you turn back the clock on one machine (e.g. reboot without fetching the current time from some server) it won't be able to connect anymore.

This may be a bit of a fundamental issue for using a client on an embedded device that has potentially no RTC, and checking the code it is indeed mentioned on the comment above here, I'm afraid that the "increasing value" here is meant to persist across restarts, and indeed what I'm observing is that at every restart it takes longer for the connection to restart, packets are dropped.

So `sys_clock_tick_get` may not work out quite well here I'm afraid. I suppose one could use an RTC device if available, or maybe this should be integrated with sntp?

jukkar replied:
Indeed the timestamp handling is a bit challenging atm. So far I did not try to tackle this problem yet and used `sys_clock_tick_get` just to get things getting forward. Using SNTP might be a good alternative here, we could initially go with that route to keep things simple.

There is the reverse problem too where we should keep track of peer timestamps. The API lets the application to configure it here https://github.com/zephyrproject-rtos/zephyr/blob/37a6185c3099379fbe47055ee16c9baacaa1356e/include/zephyr/net/wireguard.h#L71 and there is the todo item in https://github.com/zephyrproject-rtos/zephyr/blob/37a6185c3099379fbe47055ee16c9baacaa1356e/subsys/net/lib/wireguard/wg_crypto.c#L547 in which there should/could be a way to let the app to store the peer timestamp somewhere so that we could compare the peer value ourselves for replay detection.

fabiobaltieri replied:
I see, tricky, maybe just have a lightweight api to let the application override these two for now? And then maybe a followup could be to have some generic snmp or rtc subsystem integration.

jukkar replied:
> The peer side is trickier though the Linux implementation does not preserve it on reboots either so maybe it's not a problem?

Yeah, this is probably not a big issue. Everything seems to work ok without this implemented, anyway something to consider if that handshake replay avoidance is needed.

jukkar committed:
doc: Add WireGuard VPN licensing info to license file

Add WireGuard VPN licensing information to LICENSE.rst file
in the documentation.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wg: Make getting current time extensible

Allow user to provide a function that will need to get
the current time from a RTC or SNTP or similar.
Wireguard handshake replay prevention needs a monotonic
time so the application should get it from somewhere.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: wg: Support RTC with native_sim board

If running wg in native-sim, use the host clock to get the
current time. This helps to have a proper handshake when
connecting even after restarting the zephyr.exe process.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Added support for getting time from RTC, added example for native-sim how to do it. (cc: @fabiobaltieri)

ceolin commented:
As discussed today in TSC, we need to review it in the security working group due its crypto use.

ceolin commented:
After reviewing this PR, the Zephyr Security Committee has concerns regarding the introduction of cryptographic algorithms from untrusted sources.

In this case, the cryptographic implementations referenced are marked as alpha and explicitly stated as not ready for production use. Furthermore, we could not find any information regarding security vulnerability handling, responsible disclosure processes, or long-term maintenance commitments for these implementations. Given these factors, we cannot consider these sources trustworthy for inclusion in Zephyr.

Zephyr follows a security-conscious approach to cryptography, and we rely on PSA Crypto API as the primary cryptographic framework. This ensures that cryptographic operations are implemented in a vetted, standardized, and maintainable manner. Additionally, we aim to avoid propagating multiple cryptographic implementations within the project, as doing so increases maintenance complexity and the risk of security issues.

For missing functionalities—such as BLAKE2s—we should aim to have them implemented within PSA Crypto rather than introducing standalone implementations. This approach aligns with Zephyr’s security and maintainability principles.

Given these concerns, we cannot accept this PR in its current form. We encourage adapting the implementation to use PSA Crypto where possible.

If you need guidance on how to proceed with PSA integration or handling missing features, we would be happy to provide further input.

jukkar commented:
Moving this back to draft as I do not have time to work on this further. Lets see if there are developers in community that could provide acceptable crypto code.

jukkar commented:
@ceolin couple of notes
* As there was no PSA support for some of these algorithms, I thought that isolating the implementations to Wireguard lib would avoid any possible issues for people using these for other parts of the code. That is why they were placed to the directory they are in this PR.
* As the implementations of these algorithms are several years old and seems to be used in other projects too, I would think they are more or less working properly. Probably they work better than any other fresh re-implementation.
* Could security team provide implementations for the missing features, as the security team has probably the best knowledge of this domain?

jukkar commented:
* Fixing merge conflicts, no other changes

ceolin commented:
> @ceolin couple of notes
> 
> * As there was no PSA support for some of these algorithms, I thought that isolating the implementations to Wireguard lib would avoid any possible issues for people using these for other parts of the code. That is why they were placed to the directory they are in this PR.

Yes, that was indeed good. Most of these algorithms are already available in the PSA / mbedTLS, the only one I saw that is definitely not there is blake2s. So why having duplicated functionality ?

> * As the implementations of these algorithms are several years old and seems to be used in other projects too, I would think they are more or less working properly. Probably they work better than any other fresh re-implementation.

Not necessarily, specially in cryptography. Some of these implementation state that they should not be used in production that they have alpha quality, which is very problematic at least.

> * Could security team provide implementations for the missing features, as the security team has probably the best knowledge of this domain?

For sure, we started conversation with mbedTLS / PSA folks regarding missing algorithm. Obviously since this involves another project the timeline can vary. Can we change it to use PSA for what is available and work together in a temporary solution for the missing bits ? We can't simply import another cryptography implementation to Zephyr that is not properly maintained.

jukkar commented:
> Most of these algorithms are already available in the PSA / mbedTLS, the only one I saw that is definitely not there is blake2s. So why having duplicated functionality ?

I did not want to have part of the code use PSA/mbedTLS and some part use the internal implementation. But I could consider porting the code to use the PSA and leave the blake implementation as is if that would be acceptable.

jukkar commented:
Hi, as it might take some time before PSA has support for Wireguard needed crypto, I created an external module for the Wireguard support. This way people can try it more easily and give feedback to make it work better. The module can be found here https://github.com/jukkar/zephyr-wireguard. The module contain a demo of the usage and configuration with native_sim board. Note that you would need Zephyr version (4.1.99+), specifically from commit 8e908176c773980a61a44f861941228dc91c3329 ("net: shell: iface: Print VPN public key") or later to use it.
